File: internal/domain/agentrun/task_outcome.go
package agentrun

import (
	"context"
	"encoding/json"
	"strings"
)

const (
	TaskSucceeded      = "succeeded"
	TaskIncomplete     = "incomplete"
	TaskPaused         = "paused"
	TaskFailed         = "failed"
	TaskCancelled      = "cancelled"
	TaskOutcomeUnknown = "outcome_unknown"

	CompletionVerified    = "verified"
	CompletionNotRequired = "not_required"
	CompletionUnverified  = "unverified"

	StreamRunning   = "running"
	StreamEnded     = "ended"
	StreamFailed    = "failed"
	StreamCancelled = "cancelled"

	ReasonMissingArtifact = "missing_artifact"
)

// ArtifactSnapshot is the immutable raw-file identity used for delivery
// verification. SHA256 and Bytes are always the complete original bytes,
// never paginated workspace.read text or an Office extract.
type ArtifactSnapshot struct {
	ID, Path, SHA256, ContentRef string
	Bytes                        int64
	SourceIdentity               string
}

type AcceptanceCheck struct {
	ID, Kind, ExpectedDigest string
	Required                 bool
	Predicate                json.RawMessage
}

type StepSpec struct {
	ID           string
	GoalRevision int64
	Required     bool
	DependsOn    []string
	Checks       []AcceptanceCheck
}

type CheckResult struct{ ID, Status, ReasonCode, EvidenceRef string }

type StepOutcome struct {
	StepID, AttemptID                       string
	GoalRevision                            int64
	State, ReasonCode                       string
	ReceiptRefs, EvidenceRefs, ArtifactRefs []string
	CheckResults                            []CheckResult
	Claim, VerifierRevision                 string
}

type TaskOutcome struct {
	TaskID                                     string
	GoalRevision, Version                      int64
	State, Completion, ReasonCode              string
	RequiredSteps, PassedSteps, RemainingSteps []string
	EvidenceRefs, ArtifactRefs                 []string
}

type TrustedReceipt struct {
	ID, OwnerScope, TaskID, RunID, StepID, AttemptID string
	GoalRevision                                     int64
	ToolName, ToolRevision, ArgumentsDigest          string
	State, OutputSnapshotRef, CreatedAt              string
	ArtifactRefs                                     []string
}

type VerificationScope struct {
	OwnerScope, TaskID string
	GoalRevision       int64
}

type CheckEvidence struct {
	ID, CheckID, ValidatorID, ValidatorRevision, SubjectSHA256 string
	OwnerScope, TaskID, ReceiptRef, Status, PredicateDigest    string
	GoalRevision                                               int64
}

type ReceiptResolver interface {
	ResolveReceipt(context.Context, VerificationScope, string) (TrustedReceipt, error)
	ResolveArtifact(context.Context, VerificationScope, string) (ArtifactSnapshot, error)
	ResolveCheckEvidence(context.Context, VerificationScope, string) (CheckEvidence, error)
}

type TaskOutcomeView struct {
	Spinner, GreenComplete, ReplyEnded, TaskComplete bool
	Tone, Label                                      string
}

type CompletedEvent struct {
	TaskID         string
	GoalRevision   int64
	OutcomeVersion int64
	Outcome        TaskOutcome
}

type OutcomeLedger struct {
	Current TaskOutcome
	History []TaskOutcome
}

func ApplyCompletedEvent(led OutcomeLedger, ev CompletedEvent) OutcomeLedger {
	if led.Current.TaskID != "" && ev.GoalRevision < led.Current.GoalRevision {
		led.History = append(led.History, ev.Outcome)
		return led
	}
	if led.Current.TaskID != "" && ev.GoalRevision == led.Current.GoalRevision && ev.OutcomeVersion <= led.Current.Version {
		return led
	}
	if led.Current.TaskID != "" && ev.GoalRevision > led.Current.GoalRevision {
		led.History = append(led.History, led.Current)
	}
	out := ev.Outcome
	if out.TaskID == "" {
		out.TaskID = ev.TaskID
	}
	out.GoalRevision = ev.GoalRevision
	out.Version = ev.OutcomeVersion
	led.Current = out
	return led
}

func EvaluateTask(ctx context.Context, scope VerificationScope, specs []StepSpec, outcomes []StepOutcome, receipts ReceiptResolver) (TaskOutcome, error) {
	latest := map[string]StepOutcome{}
	for _, out := range outcomes {
		if out.GoalRevision != scope.GoalRevision {
			continue
		}
		prev, ok := latest[out.StepID]
		if !ok || out.AttemptID >= prev.AttemptID {
			latest[out.StepID] = out
		}
	}

	result := TaskOutcome{
		TaskID:       scope.TaskID,
		GoalRevision: scope.GoalRevision,
		Version:      1,
		State:        TaskIncomplete,
		Completion:   CompletionUnverified,
		ReasonCode:   ReasonMissingArtifact,
	}
	var artifactRefs []string
	for _, spec := range specs {
		if spec.GoalRevision != 0 && spec.GoalRevision != scope.GoalRevision {
			continue
		}
		if spec.Required {
			result.RequiredSteps = append(result.RequiredSteps, spec.ID)
		}
		out, ok := latest[spec.ID]
		passed, refs := stepPassed(ctx, scope, spec, out, ok, receipts)
		artifactRefs = append(artifactRefs, refs...)
		if !spec.Required {
			continue
		}
		if passed {
			result.PassedSteps = append(result.PassedSteps, spec.ID)
			continue
		}
		result.RemainingSteps = append(result.RemainingSteps, spec.ID)
	}
	result.ArtifactRefs = artifactRefs
	if len(result.RemainingSteps) == 0 && len(result.RequiredSteps) > 0 {
		result.State = TaskSucceeded
		result.Completion = CompletionVerified
		result.ReasonCode = ""
	}
	return result, nil
}

func stepPassed(ctx context.Context, scope VerificationScope, spec StepSpec, out StepOutcome, ok bool, receipts ReceiptResolver) (bool, []string) {
	if !ok || out.State != "succeeded" && out.State != TaskSucceeded {
		return false, nil
	}
	if out.State == "skipped" {
		return false, out.ArtifactRefs
	}
	refs := append([]string{}, out.ArtifactRefs...)
	needsArtifact := false
	var l0Check *AcceptanceCheck
	for i := range spec.Checks {
		check := spec.Checks[i]
		if !check.Required {
			continue
		}
		if check.Kind == "artifact" {
			needsArtifact = true
		}
		if check.Kind == "l0" || check.Kind == "receipt" {
			c := check
			l0Check = &c
		}
	}
	if l0Check != nil {
		if len(out.ReceiptRefs) == 0 {
			return false, refs
		}
		for _, ref := range out.ReceiptRefs {
			if receipts == nil {
				return false, refs
			}
			rec, err := receipts.ResolveReceipt(ctx, scope, ref)
			if err != nil {
				return false, refs
			}
			if rec.OwnerScope != "" && rec.OwnerScope != scope.OwnerScope {
				return false, refs
			}
			if rec.TaskID != "" && rec.TaskID != scope.TaskID {
				return false, refs
			}
			if rec.GoalRevision != 0 && rec.GoalRevision != scope.GoalRevision {
				return false, refs
			}
			if l0Check.ExpectedDigest != "" && rec.ArgumentsDigest != "" && rec.ArgumentsDigest != l0Check.ExpectedDigest {
				return false, refs
			}
		}
	}
	if receipts != nil {
		for _, evRef := range out.EvidenceRefs {
			ev, err := receipts.ResolveCheckEvidence(ctx, scope, evRef)
			if err == nil && ev.ValidatorID == "model" {
				return false, refs
			}
		}
	}
	if !needsArtifact {
		return true, refs
	}
	if len(refs) == 0 {
		return false, refs
	}
	for _, ref := range refs {
		if strings.TrimSpace(ref) == "" {
			return false, refs
		}
		art, err := receipts.ResolveArtifact(ctx, scope, ref)
		if err != nil || strings.TrimSpace(art.Path) == "" {
			return false, refs
		}
	}
	return true, refs
}

func PresentTaskOutcome(outcome TaskOutcome, stream string, artifactPaths []string) TaskOutcomeView {
	view := TaskOutcomeView{Tone: "partial", Label: "部分完成"}
	switch stream {
	case StreamEnded, StreamFailed, StreamCancelled:
		view.Spinner = false
		view.ReplyEnded = true
	case StreamRunning:
		view.Spinner = true
	default:
		view.Spinner = stream != "" && stream != StreamEnded
	}
	if stream == StreamFailed {
		view.Tone = "failed"
		view.Label = "失败"
	}
	if hasMissingArtifact(outcome, artifactPaths) {
		view.GreenComplete = false
		view.TaskComplete = false
		if outcome.State == TaskFailed {
			view.Tone = "failed"
			view.Label = "失败"
			return view
		}
		view.Tone = "partial"
		view.Label = "部分完成"
		return view
	}
	if outcome.State == TaskSucceeded && outcome.Completion == CompletionVerified && len(outcome.RemainingSteps) == 0 {
		view.Tone = "complete"
		view.Label = "已验证完成"
		view.GreenComplete = true
		view.TaskComplete = true
		return view
	}
	if outcome.State == TaskSucceeded && outcome.Completion == CompletionNotRequired {
		view.Tone = "replied"
		view.Label = "已回复"
		return view
	}
	if outcome.State == TaskFailed {
		view.Tone = "failed"
		view.Label = "失败"
	}
	return view
}

func hasMissingArtifact(outcome TaskOutcome, artifactPaths []string) bool {
	hasPath := false
	for _, path := range artifactPaths {
		if strings.TrimSpace(path) == "" {
			return true
		}
		hasPath = true
	}
	for _, ref := range outcome.ArtifactRefs {
		if strings.TrimSpace(ref) == "" {
			return true
		}
	}
	if len(outcome.RemainingSteps) > 0 {
		return true
	}
	if outcome.State == TaskSucceeded && outcome.Completion == CompletionVerified && !hasPath {
		return true
	}
	return false
}

File: web/src/session/taskOutcome.ts
export type TaskOutcomeState = 'succeeded' | 'incomplete' | 'paused' | 'failed' | 'cancelled' | 'outcome_unknown'
export type TaskCompletion = 'verified' | 'not_required' | 'unverified'

export type TaskOutcome = {
  taskId: string
  goalRevision: number
  version: number
  state: TaskOutcomeState
  completion: TaskCompletion
  reasonCode: string
  requiredSteps: string[]
  passedSteps: string[]
  remainingSteps: string[]
  evidenceRefs: string[]
  artifactRefs: string[]
}

export type TaskOutcomeEvent = {
  taskId: string
  goalRevision: number
  outcomeVersion: number
  lastSequence: number
  outcome: TaskOutcome
}

export type TaskOutcomeStore = {
  outcome?: TaskOutcome
  lastSequence?: number
  needsGet: boolean
  ignored?: 'duplicate' | 'reverse'
}

export type TaskOutcomeView = {
  spinner: boolean
  greenComplete: boolean
  replyEnded: boolean
  taskComplete: boolean
  tone: 'complete' | 'partial' | 'failed' | 'unverified' | 'replied' | 'paused' | 'running'
  label: string
}

export type StreamStatus = 'idle' | 'streaming' | 'done' | 'failed' | 'cancelled'

function missingArtifact(outcome: TaskOutcome | undefined, artifactPaths: string[] | undefined, requiredFiles: boolean | undefined): boolean {
  const paths = artifactPaths ?? []
  if (paths.some(path => !String(path).trim())) return true
  if (outcome?.artifactRefs.some(ref => !String(ref).trim())) return true
  if ((outcome?.remainingSteps.length ?? 0) > 0) return true
  const hasPath = paths.some(path => String(path).trim())
  if (requiredFiles && !hasPath) return true
  if (outcome?.state === 'succeeded' && outcome.completion === 'verified' && !hasPath) return true
  return false
}

export function presentTaskOutcome(opts: {
  outcome?: TaskOutcome
  streamStatus: StreamStatus
  artifactPaths?: string[]
  requiredFiles?: boolean
}): TaskOutcomeView {
  const ended = opts.streamStatus === 'done' || opts.streamStatus === 'failed' || opts.streamStatus === 'cancelled'
  const spinner = opts.streamStatus === 'streaming'
  const missing = missingArtifact(opts.outcome, opts.artifactPaths, opts.requiredFiles)
  if (!opts.outcome) {
    if (opts.streamStatus === 'streaming') {
      return { spinner: true, greenComplete: false, replyEnded: false, taskComplete: false, tone: 'running', label: '进行中' }
    }
    return {
      spinner: false,
      greenComplete: false,
      replyEnded: ended,
      taskComplete: false,
      tone: 'unverified',
      label: '历史结果未验证',
    }
  }
  if (opts.streamStatus === 'failed' || opts.outcome.state === 'failed') {
    return { spinner: false, greenComplete: false, replyEnded: ended || opts.outcome.state === 'failed', taskComplete: false, tone: 'failed', label: '失败' }
  }
  if (missing) {
    return { spinner: false, greenComplete: false, replyEnded: ended, taskComplete: false, tone: 'partial', label: '部分完成' }
  }
  if (opts.outcome.state === 'paused') {
    return { spinner: false, greenComplete: false, replyEnded: ended, taskComplete: false, tone: 'paused', label: '已暂停' }
  }
  if (opts.outcome.state === 'succeeded' && opts.outcome.completion === 'verified' && opts.outcome.remainingSteps.length === 0) {
    return { spinner: false, greenComplete: true, replyEnded: ended, taskComplete: true, tone: 'complete', label: '已验证完成' }
  }
  if (opts.outcome.state === 'succeeded' && opts.outcome.completion === 'not_required') {
    return { spinner: false, greenComplete: false, replyEnded: true, taskComplete: false, tone: 'replied', label: '已回复' }
  }
  return { spinner, greenComplete: false, replyEnded: ended, taskComplete: false, tone: spinner ? 'running' : 'partial', label: spinner ? '进行中' : '部分完成' }
}

export function applyTaskOutcomeEvent(prev: TaskOutcomeStore | undefined, event: TaskOutcomeEvent): TaskOutcomeStore {
  const current = prev?.outcome
  if (current) {
    if (event.goalRevision === current.goalRevision && event.outcomeVersion === current.version) {
      return { ...prev, ignored: 'duplicate', needsGet: false }
    }
    if (event.goalRevision < current.goalRevision || (event.goalRevision === current.goalRevision && event.outcomeVersion < current.version)) {
      return { ...prev, ignored: 'reverse', needsGet: false }
    }
  }
  const gap = prev?.lastSequence != null && event.lastSequence > prev.lastSequence + 1
  return { outcome: event.outcome, lastSequence: event.lastSequence, needsGet: gap }
}

export function isTaskOutcome(value: unknown): value is TaskOutcome {
  if (!value || typeof value !== 'object') return false
  const v = value as Record<string, unknown>
  return typeof v.taskId === 'string' && Number.isInteger(v.goalRevision) && Number.isInteger(v.version) && typeof v.state === 'string' && typeof v.completion === 'string'
}

export function coerceTaskOutcome(value: unknown): TaskOutcome | undefined {
  if (!isTaskOutcome(value)) return
  const v = value as TaskOutcome & { reasonCode?: string; requiredSteps?: string[]; passedSteps?: string[]; remainingSteps?: string[]; evidenceRefs?: string[]; artifactRefs?: string[] }
  return {
    taskId: v.taskId,
    goalRevision: v.goalRevision,
    version: v.version,
    state: v.state,
    completion: v.completion,
    reasonCode: v.reasonCode ?? '',
    requiredSteps: v.requiredSteps ?? [],
    passedSteps: v.passedSteps ?? [],
    remainingSteps: v.remainingSteps ?? [],
    evidenceRefs: v.evidenceRefs ?? [],
    artifactRefs: v.artifactRefs ?? [],
  }
}
